Gem/Code/Source/Components/Multiplayer/WeaponEffectComponent.cpp

```diff
@@ -0,0 +1,126 @@
+    {
+    }
+
+    void WeaponEffectComponent::OnDeactivate([[maybe_unused]] Multiplayer::EntityIsMigrating entityIsMigrating)
+    {
+    }
+
+    void WeaponEffectComponent::HandleRPC_ConfirmedHit([[maybe_unused]] AzNetworking::IConnection* invokingConnection,
+        [[maybe_unused]] const AZ::Vector3& hitPosition)
+    {
+        // Don't play the effect on the originating player as the weapon effect was predicated there already.
+        if (IsNetEntityRoleAutonomous() == false)
+        {
+            // Note: ideally, the animation needs to be played to the frame of shooting before grabbing the bone position.
+            const AZ::Vector3 start = GetNetworkWeaponsComponent()->GetCurrentShotStartPosition();
+            PlayParticleEffect(start, hitPosition);
+        }
+    }
+
+    void WeaponEffectComponent::PlayParticleEffect(const AZ::Vector3& start, const AZ::Vector3& end)
+    {
+        if (PopcornFX::PopcornFXEmitterComponentRequests* particle =
+            PopcornFX::PopcornFXEmitterComponentRequestBus::FindFirstHandler(GetEntityId()))
+        {
+            AZ::s32 attributeId = particle->GetAttributeId("Start");
+            if (attributeId >= 0)
+            {
+                particle->SetAttributeAsFloat3(attributeId, start);
+            }
+
+            attributeId = particle->GetAttributeId("End");
+            if (attributeId >= 0)
+            {
+                particle->SetAttributeAsFloat3(attributeId, end);
+            }
+
+            particle->Restart(true);
+        }
+    }
+
+
+    WeaponEffectComponentController::WeaponEffectComponentController(WeaponEffectComponent& parent)
+        : WeaponEffectComponentControllerBase(parent)
+    {
+    }
+
+    void WeaponEffectComponentController::OnActivate([[maybe_unused]] Multiplayer::EntityIsMigrating entityIsMigrating)
+    {
+        if (IsNetEntityRoleAutonomous() || IsNetEntityRoleAuthority())
+        {
+            GetParent().GetNetworkWeaponsComponent()->AddOnWeaponActivateEventHandler(m_weaponActivateHandler);
+            GetParent().GetNetworkWeaponsComponent()->AddOnWeaponPredictHitEventHandler(m_weaponPredictHandler);
+            GetParent().GetNetworkWeaponsComponent()->AddOnWeaponConfirmHitEventHandler(m_weaponConfirmHandler);
+        }
+    }
+
+    void WeaponEffectComponentController::OnDeactivate([[maybe_unused]] Multiplayer::EntityIsMigrating entityIsMigrating)
+    {
+        m_weaponActivateHandler.Disconnect();
+        m_weaponPredictHandler.Disconnect();
+        m_weaponConfirmHandler.Disconnect();
+    }
+
+    void WeaponEffectComponentController::OnWeaponActivate(const WeaponActivationInfo& info)
+    {
+        // This is the locally predicated effect on the player that initiated the weapon.
+        const AZ::Vector3& start = info.m_activateEvent.m_initialTransform.GetTranslation();
+        const AZ::Vector3& end = info.m_activateEvent.m_targetPosition;
+        GetParent().PlayParticleEffect(start, end);
+    }
```

Comment:
Possibly, however OnWeaponPredictHit  doesnt provide the trial info. It just tells one where the shot hit not where it came from, and I wanted to add some trail effects.

Reply:
I left a comment on this here:
```
    void WeaponEffectComponentController::OnWeaponPredictHit([[maybe_unused]] const WeaponHitInfo& info)
    {
        /*
         * At the moment, the particle effect plays both the trace and the hit effect in one go when predicted on the originated player.
         * If the particle effects were separated, one could spawn an entity to play the hit effect only.
         * Then this call would play the predicted hit effect, while @OnWeaponActivate would only play the trace line effect.
         */
    }
```

Reply:
I did see that. Maybe it makes sense to ad that info to Predict. My main concern is this would cause two effects to play.

Reply:
does OnWeaponPredictHit get called if there is no hit, e.g. if the player shoots into the sky?

Reply:
Last I saw, the raycast for shooting will search within a fixed distance for a target. If none is found it will end at the extent of that distance. If that hasn't changed, it should still work.

Reply:
I can confirm that `OnWeaponActivate` only gets called once on the originating client, only one on a server and only once on other clients.

Reply:
I did a cleanup pass on this. It should be good.

Reply:
> does OnWeaponPredictHit get called if there is no hit, e.g. if the player shoots into the sky?

@AMZN-alexpete no, it doesn't. If there was not hit, no predict call is called.